pootle/apps/pootle_store/diff.py

```diff
@@ -0,0 +1,306 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+from collections import OrderedDict
+import difflib
+
+from django.utils.functional import cached_property
+
+from .util import OBSOLETE
+from .fields import to_python as multistring_to_python
+
+
+class UnitDiff(object):
+
+    def __init__(self, db_unit, file_unit):
+        self.db_unit = db_unit
+        self.file_unit = file_unit
+
+    def has_changes(self):
+        if self.context_changed:
+            return True
+        if self.source_changed:
+            return True
+        if self.target_changed:
+            return True
+        if self.developer_comment_changed:
+            return True
+        if self.translator_comment_changed:
+            return True
+        if self.locations_changed:
+            return True
+
+    @property
+    def db_context(self):
+        return self.db_unit["developer_comment"] or ""
+
+    @property
+    def db_developer_comment(self):
+        return self.db_unit["developer_comment"] or ""
+
+    @property
+    def db_translator_comment(self):
+        return self.db_unit["translator_comment"] or ""
+
+    @property
+    def db_source(self):
+        return multistring_to_python(self.db_unit["source_f"])
+
+    @property
+    def db_target(self):
+        return multistring_to_python(self.db_unit["target_f"])
+
+    @property
+    def context_changed(self):
+        return self.db_context != self.file_unit["context"]
+
+    @property
+    def developer_comment_changed(self):
+        return self.db_developer_comment != self.file_unit["developer_comment"]
+
+    @property
+    def file_locations(self):
+        return self.db_unit["locations"].split("\n")
+
+    @property
+    def locations_changed(self):
+        return self.db_unit["locations"] != self.file_locations
+
+    @property
+    def source_changed(self):
+        return not ((self.file_unit["source"] == "" and self.db_source == "")
```

Comment:
@ta2-1 the one remaining concern i have is whether this test is exactly correct for detecting db/file unit differences - similar for target_changed